crates/data-structures/src/slim_slice.rs

```diff
@@ -0,0 +1,1934 @@
+    type Error = LenTooLong<Box<[T]>>;
+
+    #[inline]
+    fn try_from(boxed: Box<[T]>) -> Result<Self, Self::Error> {
+        ensure_len_fits!(boxed);
+        // SAFETY: Checked above that `len <= u32::MAX`.
+        Ok(unsafe { Self::from_boxed_unchecked(boxed) })
+    }
+}
+
+impl<T> TryFrom<Vec<T>> for SlimSliceBox<T> {
+    type Error = LenTooLong<Vec<T>>;
+
+    #[inline]
+    fn try_from(vec: Vec<T>) -> Result<Self, Self::Error> {
+        ensure_len_fits!(vec);
+        // SAFETY: Checked above that `len <= u32::MAX`.
+        Ok(unsafe { Self::from_boxed_unchecked(vec.into_boxed_slice()) })
+    }
+}
+
+impl<T, const N: usize> From<[T; N]> for SlimSliceBox<T> {
+    #[inline]
+    fn from(arr: [T; N]) -> Self {
+        #[allow(clippy::let_unit_value)]
+        let () = AssertU32::<N>::OK;
+
+        // SAFETY: We verified statically by `AssertU32<N>` above that `N` fits in u32.
+        unsafe { Self::from_boxed_unchecked(into_box(arr)) }
+    }
+}
+
+impl<T> From<SlimSliceBox<T>> for Vec<T> {
+    #[inline]
+    fn from(slice: SlimSliceBox<T>) -> Self {
+        into_box(slice).into()
+    }
+}
+
+impl<T: Debug> Debug for SlimSliceBox<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(self.deref(), f)
+    }
+}
+
+impl<T: Clone> Clone for SlimSliceBox<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        // Allocate exactly the right amount
+        // so we later don't reallocate due to excess capacity.
+        let mut vec = Vec::with_capacity(self.len());
+        vec.extend_from_slice(self);
+        // SAFETY: We know `self.len() <= u32::MAX`.
+        unsafe { Self::from_boxed_unchecked(into_box(vec)) }
+    }
+}
+
+impl<R: Deref, T> PartialEq<R> for SlimSliceBox<T>
+where
+    [T]: PartialEq<R::Target>,
+{
+    #[inline]
+    fn eq(&self, other: &R) -> bool {
+        **self == **other
+    }
+}
+
+impl<T: Eq> Eq for SlimSliceBox<T> {}
+
+impl<R: Deref, T> PartialOrd<R> for SlimSliceBox<T>
+where
+    [T]: PartialOrd<R::Target>,
+{
+    #[inline]
+    fn partial_cmp(&self, other: &R) -> Option<Ordering> {
+        (**self).partial_cmp(&**other)
+    }
+}
+
+impl<T: Ord> Ord for SlimSliceBox<T> {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        (**self).cmp(&**other)
+    }
+}
+
+impl<T: Hash> Hash for SlimSliceBox<T> {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Hash::hash(&**self, state)
+    }
+}
+
+impl<T> IntoIterator for SlimSliceBox<T> {
+    type Item = T;
+    type IntoIter = <Vec<T> as IntoIterator>::IntoIter;
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        Vec::from(self).into_iter()
+    }
+}
+
+/// A wrapper to achieve `FromIterator<T> for Result<SlimSliceBox<T>, Vec<T>>`.
+///
+/// We cannot do this directly due to orphan rules.
+pub struct SlimSliceBoxCollected<T> {
+    /// The result of `from_iter`.
+    pub inner: Result<SlimSliceBox<T>, LenTooLong<Vec<T>>>,
```

Comment:
This feels... inelegant. Maybe we could have a `try_from_iter() -> Result<Self, LenTooLong<Vec<T>>>` function on SlimSliceBox + a `try_collect_slim_slice_box(self)` extension trait method? Idk. Not blocking this on a better solution, this is fine, but just feels like it could be better.